### PR TITLE
rust - update to version 1.64.0

### PR DESCRIPTION
**Pull request by fritzkink**
currently there is no system/library/g++-4-runtime installed only two links are needed
to build the package:

      `ln -s $(GCC_LIBDIR)/libgcc_s.so.1 /usr/lib/$(MACH64)/libgcc_s.so.1`
      `ln -s $(GCC_LIBDIR)/liblibssp.so.0 /usr/lib/$(MACH64)/libssp.so.0`

 another idea could be to usr LD_LIBRARY_PATH only caveat it starts building included
 llvm and fails (maybe someone can chime in)

      `export LD_LIBRARY_PATH=$(GCC_LIBDIR)`

 Also some other package can provide this two links (which one?)
 Basically it proves that the very old gcc is not needed as also mentioned by Joshua M. Clulow

### CODE REVIEW

**Comment by AndWac** [comment]
This doesn't build here:

   Compiling toml v0.5.9
    Finished dev [unoptimized] target(s) in 41.40s
Building stage0 std artifacts (x86_64-unknown-illumos -> x86_64-unknown-illumos)
error: unknown `-Z` flag specified: check-cfg
Build completed unsuccessfully in 0:00:42
make[1]: *** [Makefile:54: install] Error 1
make[1]: Leaving directory '/export2/home/andreas/oi-userland/components/developer/rust/build/amd64'
gmake: *** [/export2/home/andreas/oi-userland/make-rules/configure.mk:183: /export2/home/andreas/oi-userland/components/developer/rust/build/amd64/.installed] Error 2

**Comment by fritzkink** [comment]
Hmm, I just run the build target again on my system  without any problem. I also don't see where a -z flag is used in the Makefile.

**Comment by AndWac** [comment]
I have checked this on my 2nd build system with the same error. Are you sure your system is up-to-date and clean? I cannot merge this because it is most likely that this will also fail on the official build system.

**Comment by fritzkink** [comment]
It seems that all settings for the COMPONENT_BUILD_ENV from the Makefile are ignored. Did something change here? Basically my environment is up to date.

**Comment by AndWac** [comment]
@mtelka: Do you have an idea? It would break many package builds if that is generally true.

**Comment by mtelka** [comment]
> It seems that all settings for the COMPONENT_BUILD_ENV from the Makefile are ignored.

@fritzkink: Why do you think this happens?  Any steps and/or logs supporting your claim?  Thanks.

**Comment by mtelka** [comment]
I just tried to rebase [fritzkink:rust](https://github.com/fritzkink/oi-userland/tree/rust) on top of current git head (with all packages up-to-date) and then run `gmake sample-manifest` and this is the result (the full path was manually replaced by `XXXYYYXXX` in the output):

```
...
...
/usr/bin/touch XXXYYYXXX/components/developer/rust/build/amd64/.built
mkdir -p XXXYYYXXX/components/developer/rust/build/prototype/i386
(cd XXXYYYXXX/components/developer/rust/build/amd64 ; /usr/bin/env OPENSSL_DIR=/usr/openssl/1.1 OPENSSL_LIB_DIR=/usr/openssl/1.1/lib/amd64 OPENSSL_INCLUDE_DIR=/usr/openssl/1.1/include OPENSSL_STATIC=0 CC=/usr/gcc/7/bin/gcc CFLAGS="-m64 -O3  -D_POSIX_PTHREAD_SEMANTICS" CXX=/usr/gcc/7/bin/g++ CXXFLAGS="-m64 -O3  -D_POSIX_PTHREAD_SEMANTICS" AR=/usr/gnu/bin/ar RUSTC=XXXYYYXXX/components/developer/rust/build/amd64/bin/rustc MAKE=/usr/bin/python3.9 RUSTFLAGS="-C linker=/usr/gcc/7/bin/gcc" RUST_BACKTRACE=1 DESTDIR=XXXYYYXXX/components/developer/rust/build/prototype/i386 /usr/gnu/bin/make \
                ./x.py  install)
make[1]: Entering directory 'XXXYYYXXX/components/developer/rust/build/amd64'
make[1]: Nothing to be done for 'x.py'.
Building rustbuild
error: rustup could not choose a version of cargo to run, because one wasn't specified explicitly, and no default is configured.
help: run 'rustup default stable' to download the latest stable release of Rust and set it as your default toolchain.
failed to run: XXXYYYXXX/components/developer/rust/build/amd64/bin/cargo build --manifest-path XXXYYYXXX/components/developer/rust/rustc-1.64.0-src/src/bootstrap/Cargo.toml --frozen
Build completed unsuccessfully in 0:00:00
make[1]: *** [Makefile:54: install] Error 1
```

To make it short: it does not install.  Why?  I do not know and I'd prefer to not go to the rust hell yet if possible.  For now Python hell is enough for me. :-)

File: components/developer/rust/Makefile [hunk 1]
```diff
@@ -64,7 +64,7 @@ CARGO_HOME=             $(BUILD_DIR)/$(MACH64)
 
 PATH=                   $(CARGO_HOME)/bin:$(GCC_BINDIR):$(PATH.gnu)
 
-#  workaround the symlink name lenght 100 problem, but hope it is not neccessary with
+# workaround the symlink name lenght 100 problem, but hope it is not neccessary with
 # but also with patch src_tools_rust-installer_src_generator.rs.patch 
 COMPONENT_BUILD_ENV+=	CARGO_HOME=$(CARGO_HOME)
```
Comment (mtelka) [inline components/developer/rust/Makefile:69]: There are some strange things in the `Makefile`.  Something is added to the `COMPONENT_BUILD_ENV` here...

File: components/developer/rust/Makefile [hunk 2]
```diff
@@ -120,19 +134,18 @@ CONFIGURE_OPTIONS += --enable-rpath
...
 CONFIGURE_OPTIONS += --enable-llvm-link-shared
 CONFIGURE_OPTIONS += --llvm-root=$(CLANG_ROOT)
-
-COMPONENT_PRE_BUILD_ACTION=	rustup default  $(RUST_PRE_VERSION)-$(RUST_ARCH);
+CONFIGURE_OPTIONS += --enable-llvm-version-check
+CONFIGURE_OPTIONS += --enable-optimize-llvm
+CONFIGURE_OPTIONS += --enable-control-flow-guard
 
 # Cleanup standard environment!
 COMPONENT_BUILD_ENV =
```
Comment (mtelka) [inline components/developer/rust/Makefile:151]: ...and then the `COMPONENT_BUILD_ENV` is cleaned here.

Definitely, this `Makefile` needs some love.

**Review by mtelka — CHANGES_REQUESTED** [review]

**Comment by fritzkink** [comment]
superseeded by a newer pull request
